File: javascript/day14_part2_2017.js
const fs = require('fs');

function reverseSection(arr, start, length) {
    const n = arr.length;
    for (let i = start, j = start + length - 1; i < j; i++, j--) {
        [arr[i % n], arr[j % n]] = [arr[j % n], arr[i % n]];
    }
}

function knotHash(input) {
    let lengths = [];
    for (let char of input) {
        lengths.push(char.charCodeAt(0));
    }
    lengths.push(17, 31, 73, 47, 23);

    let list = Array.from({ length: 256 }, (_, i) => i);

    let position = 0;
    let skip = 0;
    for (let round = 0; round < 64; round++) {
        for (let length of lengths) {
            reverseSection(list, position, length);
            position += length + skip;
            skip++;
        }
    }

    let denseHash = [];
    for (let i = 0; i < 16; i++) {
        let xor = 0;
        for (let j = 0; j < 16; j++) {
            xor ^= list[i * 16 + j];
        }
        denseHash.push(xor);
    }

    let hexHash = denseHash.map(v => v.toString(16).padStart(2, '0')).join('');
    return hexHash;
}

function hexToBinary(hexStr) {
    let binaryStr = '';
    for (let hexDigit of hexStr) {
        let val = parseInt(hexDigit, 16);
        binaryStr += val.toString(2).padStart(4, '0');
    }
    return binaryStr;
}

function dfs(x, y, grid) {
    if (x < 0 || x >= 128 || y < 0 || y >= 128 || grid[x][y] !== 1) {
        return;
    }
    grid[x][y] = 0;
    dfs(x - 1, y, grid);
    dfs(x + 1, y, grid);
    dfs(x, y - 1, grid);
    dfs(x, y + 1, grid);
}

const data = fs.readFileSync('input.txt', 'utf8').trim();
const grid = Array.from({ length: 128 }, () => Array(128).fill(0));
let totalUsed = 0;
let regions = 0;

for (let i = 0; i < 128; i++) {
    const rowKey = `${data}-${i}`;
    const hash = knotHash(rowKey);
    const binaryRow = hexToBinary(hash);

    for (let j = 0; j < 128; j++) {
        if (binaryRow[j] === '1') {
            grid[i][j] = 1;
            totalUsed++;
        }
    }
}

for (let i = 0; i < 128; i++) {
    for (let j = 0; j < 128; j++) {
        if (grid[i][j] === 1) {
            regions++;
            dfs(i, j, grid);
        }
    }
}

console.log(regions);

File: csharp/day14_part2_2017.cs

using System;
using System.IO;
using System.Text;

class Program
{
    static void ReverseSection(int[] arr, int start, int length)
    {
        int n = arr.Length;
        for (int i = start, j = start + length - 1; i < j; i++, j--)
        {
            int tmp = arr[i % n];
            arr[i % n] = arr[j % n];
            arr[j % n] = tmp;
        }
    }

    static string KnotHash(string input)
    {
        int[] lengths = new int[input.Length];
        for (int i = 0; i < input.Length; i++)
        {
            lengths[i] = (int)input[i];
        }
        int[] extraLengths = { 17, 31, 73, 47, 23 };
        Array.Resize(ref lengths, lengths.Length + extraLengths.Length);
        extraLengths.CopyTo(lengths, input.Length);

        int[] list = new int[256];
        for (int i = 0; i < list.Length; i++)
        {
            list[i] = i;
        }

        int position = 0;
        int skip = 0;
        for (int round = 0; round < 64; round++)
        {
            foreach (int length in lengths)
            {
                ReverseSection(list, position, length);
                position += length + skip;
                skip++;
            }
        }

        int[] denseHash = new int[16];
        for (int i = 0; i < 16; i++)
        {
            int xor = 0;
            for (int j = 0; j < 16; j++)
            {
                xor ^= list[i * 16 + j];
            }
            denseHash[i] = xor;
        }

        StringBuilder hexHash = new StringBuilder();
        foreach (int v in denseHash)
        {
            hexHash.Append(v.ToString("x2"));
        }
        return hexHash.ToString();
    }

    static string HexToBinary(string hexStr)
    {
        StringBuilder binaryStr = new StringBuilder();
        foreach (char hexDigit in hexStr)
        {
            int val = Convert.ToInt32(hexDigit.ToString(), 16);
            binaryStr.Append(Convert.ToString(val, 2).PadLeft(4, '0'));
        }
        return binaryStr.ToString();
    }

    static void DFS(int x, int y, int[][] grid)
    {
        if (x < 0 || x >= 128 || y < 0 || y >= 128 || grid[x][y] != 1)
        {
            return;
        }
        grid[x][y] = 0;
        DFS(x - 1, y, grid);
        DFS(x + 1, y, grid);
        DFS(x, y - 1, grid);
        DFS(x, y + 1, grid);
    }

    static void Main()
    {
        string keyString = File.ReadAllText("input.txt").Trim();
        int[][] grid = new int[128][];
        int totalUsed = 0;
        int regions = 0;

        for (int i = 0; i < 128; i++)
        {
            grid[i] = new int[128];
            string rowKey = keyString + "-" + i;
            string hash = KnotHash(rowKey);
            string binaryRow = HexToBinary(hash);

            for (int j = 0; j < binaryRow.Length; j++)
            {
                if (binaryRow[j] == '1')
                {
                    grid[i][j] = 1;
                    totalUsed++;
                }
            }
        }

        for (int i = 0; i < 128; i++)
        {
            for (int j = 0; j < 128; j++)
            {
                if (grid[i][j] == 1)
                {
                    regions++;
                    DFS(i, j, grid);
                }
            }
        }

        Console.WriteLine(regions);
    }
}
